### DISCUSSION TITLE
How to get LayoutContext and Constraints to use in CalcPreferredSize()?

### DISCUSSION THREAD

**Discussion by PiotrRyw** [isl-org/Open3D#6468] [Q&A]
I don't see how to get LayoutContext and Constraints from application instance or from window instance or from anywhere, to later pass to CalcPreferredSize() of widgets.

I can manually create an instance of Constraints, but when I try to create an instance LayoutContext I get error messages such as "C++ the default constructor of open3d::visualization::gui::LayoutContext cannot be referenced -- it is a deleted function".

In the documentation of the struct LayoutContext there is no explanation of how to obtain it. GetLayoutContext() in Window class is a protected function.
Is it possible to use CalcPreferredSize?
The same goes for any other function necessitating LayoutContext, eg. CalcMinimumSize().

I've looked through your source code, as the documentation does indeed exist.

Why is CalcPreferredSize() implemented as a public function for every widget if it is impossible to access LayoutContext required by the CalcPrefferedSize(), as the only function returning LayoutContext ( open3d::visualization::gui::Window::GetLayoutContext() ) is protected and the default constructor of LayoutContext is deleted?
Am I missing something that would benefit the C++ documentation?

**Comment by EwingKang** [comment]
I've came to the same conclusion as @PiotrRyw here, I think it might be an WIP or there might be some design rethink needed.
I'm trying to change the oversized black boarder around a Widget, e.g. sliders/ Labels. And I found it impossible for my to control it using a simple [Widget::SetFrame](http://www.open3d.org/docs/latest/cpp_api/classopen3d_1_1visualization_1_1gui_1_1_widget.html#a0cff4642a394b5d2947bc38c31ac1c2e)

One side note is that I'm not able to find the virtual implementation of `Layout` for `Label` object. It is very strange that `Label` inherits `Widget` without overriding the `Layout()` function. However, in the implementation of `Layout()` of `Widget`, it actually does nothing since it only goes through the children widgets, which a single `Label` object has none
https://github.com/isl-org/Open3D/blob/a590c7774e1afef91e943950942f551b47ba76b8/cpp/open3d/visualization/gui/Label.h#L16
https://github.com/isl-org/Open3D/blob/a590c7774e1afef91e943950942f551b47ba76b8/cpp/open3d/visualization/gui/Widget.cpp#L86-L90
